File: DAYS/Day07-FindMin/findMin.py
# Array Problem : Find Minimum in Rotated Sorted Array
# Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
# Ex: [4,5,6,7,0,1,2] might become [0,1,2,4,5,6,7]
# Example 1: input: nums = [3,4,5,1,2] output: 1
# Example 2: input: nums = [4,5,6,7,0,1,2] output: 0
# Example 3: input: nums = [11,13,15,17] output: 11
# solution by Binary Search

# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            # If the middle element is greater than the rightmost element,
            # the minimum element must be in the right half.
            if nums[mid] > nums[right]:
                left = mid + 1
            # If the middle element is less than or equal to the rightmost element,
            # the minimum element must be in the left half.
            else:
                right = mid
        return nums[left]

# Test Cases
print(Solution().findMin([3,4,5,1,2]))  # Output: 1
print(Solution().findMin([4,5,6,7,0,1,2]))  # Output: 0
print(Solution().findMin([11,13,15,17]))  # Output: 11

# Best Approach: Binary Search
# Complexity Analysis: Time: O(log n) | Space: O(1)

File: DAYS/Day07-FindMin/findMin.js
// Array Problem : Find Minimum in Rotated Sorted Array
// Given the array nums after the rotation and an integer target, 
// return the index of target if it is in nums, or -1 if it is not in nums.
// Example 1: Input: nums = [4,5,6,7,0,1,2], Output: 0
// Example 2: Input: nums = [3,4,5,1,2], Output: 1
// Example 3: Input: nums = [1],  Output: 1

// Leetcode Problem Link : https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
// Solution : Using Binary Search

function findMin(nums) {
    let left=0;
    let right=nums.length-1;
    while (left < right){
        let mid = Math.floor((left +right /2));
        if(nums[mid]>nums[right]){
            left=mid +1
        }else{
            right=mid
        }
    }return nums[left];
}
console.log(findMin([4,5,6,7,0,1,2])); // Output: 0
console.log(findMin([1])); // Output: 1
console.log(findMin([3,4,5,1,2])); // Output: 1
